### PR TITLE
Encounters API & statusEffect helpers

### PR DESCRIPTION
**Pull request by aimozg**
The small changes are deprecation and replacement of `findStatusEffect():int` with `hasStatusEffect():Boolean`, `statusEffectByType():StatusEffectClass` and `indexOfStatusEffect():int` and an introduction of includeable file to inject BaseContent and Utils functionality without the need of subclassing. Old functions and classes are left for compatibility.

The other is new Encounter API, used as a demo in `Exploration.as`, to make much easier:
* adding a possible event to pool of random encounters
* make one event appear more/less often
* add a condition to the event
* apply a level- corruption- whatever-dependent encounter chance adjustment
* (maybe the most important) understand from the code, what events can happen during exploration, and what they depend on.

The configuration looks like this:
```as3
Encounters.group(game.commonEncounters.withImpGob, {
  name  : "deepwoods",
  call  : discoverDeepwoods,
  when  : function ():Boolean {
    return (flags[kFLAGS.TIMES_EXPLORED_FOREST] >= 20) 
    && !player.hasStatusEffect(StatusEffects.ExploredDeepwoods);
  },
  chance: Encounters.ALWAYS
}, {
  name  : "beegirl",
  call  : beeGirlScene.beeEncounter,
  chance: 0.50
}, {
  name: "tentabeast",
  call: tentacleBeastEncounterFn,
  when: fn.ifLevelMin(2)
}, {
  name  : "succubus",
  call  : game.succubusScene.encounterSuccubus,
  when  : fn.ifLevelMin(3),
  mods  : [game.commonEncounters.lethiteMod],
  chance: 0.50
}, /*...*/ )
```
and behind the scene it does this:
```
[trace] group{18}.execEncounter()
[trace]   selecting from common=1 tamani=0 jojo=0 essrayle=0 corrGlade=0.87 lumber=0.5 deepwoods=0 beegirl=0.5 tentabeast=1 mimic=0.5 succubus=0.5 marble=0.1 trip=1 chitin=0.05 healpill=0.1 truffle=0.35 bigjunk=0 walk=1
[trace]   -> picked encounter common with chance 1 of total 7.47
[trace]   common.execEncounter()
[trace]     selecting from imps=0.5 goblins=0.5 common=0
[trace]     -> picked encounter goblins with chance 0.5 of total 1
[trace]     goblins.execEncounter()
[trace]       selecting from goblin=0.868 gobass=0 gobwar=0 gobsha=0 gobeld=0
[trace]       -> picked encounter goblin with chance 0.868 of total 0.868
```

Level-depended encounters are configured that way:
```as3
{
  name  : "implord",
  call  : kGAMECLASS.impScene.impLordEncounter,
  when  : fn.ifLevelMin(8),
  chance: fn.lineByLevel(8, 16, 0.5, 2.0) // x0.5 at level 8, x2 at level 16+
}
```

### CODE REVIEW

**Comment by Stadler76** [comment]
```as3
{
  name  : "implord",
  call  : kGAMECLASS.impScene.impLordEncounter,
  when  : fn.ifLevelMin(8),
  chance: fn.lineByLevel(8, 16, 0.5, 2.0) // x0.5 at level 8, x2 at level 16+
}
```
shouldn't that be
```as3
  chance: fn.lineByLevel(8, 0.5, 16, 2.0) // x0.5 at level 8, x2 at level 16+
```
?

PS: The Encounters API looks interesting, but affecting 200+ files in this PR is quite much IMHO

**Comment by aimozg** [comment]
@Stadler76 
Perhaps. The Encounters are contained in `1d0edda` that can be merged separately.

Argument order in `lineByLevel`  is a matter a taste IMHO - I assume you prefer the _x1,y1,x2,y2_ as a sequence of points. I could change it if there's solid reason..
See, there's also upcoming `logScale(x1,x2,x3,y1,y3)` that returns params for _y = a * ln(x - c) + b_, and it requires _y2_ to be in the middle.

File: classes/classes/CharCreation.as [hunk 1]
```diff
@@ -1403,7 +1403,7 @@
 		private function startTheGame():void {
 			player.startingRace = player.race();
 			if (flags[kFLAGS.HARDCORE_MODE] > 0) {
-				trace("Hardcore save file " + flags[kFLAGS.HARDCORE_SLOT] + " created.")
+				trace("Hardcore save file " + flags[kFLAGS.HARDCORE_SLOT] + " created.");
 				getGame().saves.saveGame(flags[kFLAGS.HARDCORE_SLOT])
```
Comment (Wishbringer) [inline classes/classes/CharCreation.as]: Missed one here. :D

Comment (aimozg) [inline classes/classes/CharCreation.as]: Run inspection by name / Unterminated statement / Whole project
...967 warnings

**Comment by Stadler76** [comment]
It seems, that the pics for goblin and imp encounters are missing now.

**Comment by Stadler76** [comment]
Ok, the following diff should fix this: 
[20170409_0128_gobbo-imp-sprites-fix.diff.txt](https://github.com/Kitteh6660/Corruption-of-Champions-Mod/files/907878/20170409_0128_gobbo-imp-sprites-fix.diff.txt)
PS: Ok, srsly: Why doesn't GitHub accept .diff-files or files zipped with 7-zip?!?

**Comment by aimozg** [comment]
@Stadler76 Thanks for fixing and sorry for late response and that stupid `hasStatusEffect` bug.

**Pull request merged**